airflow/models/dagrun.py

```diff
@@ -344,6 +393,9 @@ def update_state(self, session: Session = None) -> List[TI]:
         leaf_task_ids = {t.task_id for t in dag.leaves}
         leaf_tis = [ti for ti in tis if ti.task_id in leaf_task_ids]
 
+        # TODO[ha]: These callbacks shouldn't run in the scheduler loop - check if Kamil changed this to run
```

Comment:
We should indeed move it outside of the scheduler loop to a DAG processor. And possibly  send it to DagFileProcessor like Kamil did. 

This shoud actually help a lot if this is all within the uncommitted transaction. Those callbacks are potentialy problematic

Reply:
Yes, this piece is ready, should hopefully be pushed today.